ethereum/statetransition/src/main/java/tech/pegasys/artemis/statetransition/util/AttestationUtil.java

```diff
@@ -309,47 +304,92 @@ public static int ceil_div8(int input) {
   }
 
   /**
-   * get total balance of validators attesting to state for the given block_root
+   * get indices of validators attesting to state for the given block_root TODO: the union part
+   * takes O(n^2) time, where n is the number of validators. OPTIMIZE
    *
    * @param state
    * @param crosslink_committee
-   * @param shard
+   * @param shard_block_root
    * @return
    * @throws BlockValidationException
    */
-  public static UnsignedLong total_attesting_balance(
+  public static List<Integer> attesting_validator_indices(
       BeaconState state, CrosslinkCommittee crosslink_committee, Bytes32 shard_block_root)
       throws Exception {
-    List<Integer> attesting_validator_indices =
-        attesting_validator_indices(state, crosslink_committee, shard_block_root);
-    return BeaconStateUtil.get_total_effective_balance(state, attesting_validator_indices);
+    UnsignedLong current_epoch = BeaconStateUtil.get_current_epoch(state);
+    UnsignedLong previous_epoch = BeaconStateUtil.get_previous_epoch(state);
+    List<PendingAttestation> combined_attestations = get_epoch_attestations(state, current_epoch);
+    combined_attestations.addAll(get_epoch_attestations(state, previous_epoch));
+
+    List<ArrayList<Integer>> validator_index_sets = new ArrayList<>();
+
+    for (PendingAttestation attestation : combined_attestations) {
+      if (attestation.getData().getShard().compareTo(crosslink_committee.getShard()) == 0
+          && attestation.getData().getShard_block_root() == shard_block_root) {
+        validator_index_sets.add(
+            BeaconStateUtil.get_attestation_participants(
+                state, attestation.getData(), attestation.getAggregation_bitfield().toArray()));
+      }
+    }
+
+    List<Integer> attesting_validator_indices = new ArrayList<Integer>();
+    for (List<Integer> validator_index_set : validator_index_sets) {
+      for (Integer validator_index : validator_index_set) {
+        if (!attesting_validator_indices.contains(validator_index)) {
```

Comment:
you should mark that call to .contains() as an O(n) operation that could be optimized